src/vs/workbench/services/runtimeSession/common/runtimeSession.ts

```diff
@@ -2019,7 +2027,123 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
+
+		// Find the session associated with the old URI
+		const session = this._notebookSessionsByNotebookUri.get(oldUri);
+
+		if (!session) {
+			// No matching session found for the provided oldUri
+			// Why logging as debug: This is an expected case when notebooks don't have sessions yet
+			this._logService.debug(`No notebook session found for URI: ${oldUri.toString()}`);
+			return undefined;
+		}
+
+		// Check if session is in a valid state for URI reassignment
+		// Why: We can't reassign a terminated session as it's no longer active and would cause
+		// users to think they have a working session when they don't
+		if (session.getRuntimeState() === RuntimeState.Exited) {
+			this._logService.warn('Cannot update URI for terminated session', {
+				sessionId: session.sessionId,
+				oldUri: oldUri.toString()
+			});
+			return undefined;
+		}
+
+		// Remember the session ID for return value
+		const sessionId = session.sessionId;
+
+		try {
+			// Operations are performed in a specific order to maintain atomic-like behavior
+			// The ordering ensures that even if interrupted between steps, the system won't lose
+			// track of the session completely.
+
+			// 1. First add the new mapping to ensure we don't lose the session
+			// Why: This makes the session accessible via the new URI immediately,
+			// so even if the next steps fail, the session is still accessible via some URI
+			this._notebookSessionsByNotebookUri.set(newUri, session);
+
+			// 2. Then update the session's notebook URI in its dynamic state
+			// Why: This ensures the session's internal references are consistent
+			// with our mapping, which helps debugging and ensures session properties
+			// reflect current reality
+			session.dynState.currentNotebookUri = newUri;
+
+			// 3. Finally remove the old mapping - we do this last because it's
+			// the most likely to fail if ResourceMap has internal inconsistency
+			// Why last: If we deleted first and then failed to add the new mapping,
+			// we'd lose the session entirely
+			this._notebookSessionsByNotebookUri.delete(oldUri);
+
+			// Log success for debugging
+			this._logService.debug(`Successfully updated notebook session URI: ${oldUri.toString()} → ${newUri.toString()}`);
+
+			// Notify listeners that the URI has been updated
+			// Why: Components like the variables view need to update their UI
+			// to show the new filename instead of "Untitled-1", and any code that tracks
+			// notebook URIs needs to update its references
+			this._onDidUpdateNotebookSessionUriEmitter.fire({
+				sessionId,
+				oldUri,
+				newUri
+			});
+
+			return sessionId;
+		} catch (error) {
+			// If anything went wrong, attempt to restore the old state manually
+			this._logService.error('Failed to update notebook session URI', error);
+
+			// Manual restoration in reverse order to maintain consistency
+			// Why: We need to undo all the changes to return to the previous state
+			try {
```

Comment:
Is this nested try/catch necessary? It doesn't look like anything in the try block could throw.

Reply:
Good catch. Fixed